EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/AccountLinkManager.java

```diff
@@ -0,0 +1,110 @@
+package net.essentialsx.discordlink;
+
+import com.earth2me.essentials.IEssentialsModule;
+import net.ess3.api.IUser;
+import net.essentialsx.api.v2.services.discord.InteractionMember;
+import net.essentialsx.discord.JDADiscordService;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class AccountLinkManager implements IEssentialsModule {
+    private static final char[] CODE_CHARACTERS = "abcdefghijklmnopqrstuvwxyz0123456789".toCharArray();
+
+    private final EssentialsDiscordLink ess;
+    private final AccountStorage storage;
+
+    private final Map<String, UUID> codeToUuidMap = new ConcurrentHashMap<>();
+
+    public AccountLinkManager(EssentialsDiscordLink ess, AccountStorage storage) {
+        this.ess = ess;
+        this.storage = storage;
+    }
+
+    public boolean isLinked(final UUID uuid) {
+        return getDiscordId(uuid) != null;
+    }
+
+    public boolean isLinked(final String discordId) {
+        return getUUID(discordId) != null;
+    }
+
+    public String createCode(final UUID uuid) throws IllegalArgumentException {
+        synchronized (codeToUuidMap) {
+            final Optional<Map.Entry<String, UUID>> prevCode = codeToUuidMap.entrySet().stream().filter(stringUUIDEntry -> stringUUIDEntry.getValue().equals(uuid)).findFirst();
+            if (prevCode.isPresent()) {
+                throw new IllegalArgumentException(prevCode.get().getKey());
+            }
+
+            final String code = generateCode();
+
+            codeToUuidMap.put(code, uuid);
+            return code;
+        }
+    }
+
+    public UUID getPendingUUID(final String code) {
+        synchronized (codeToUuidMap) {
+            return codeToUuidMap.remove(code);
```

Comment:
Would prefer this to be a pure getter; removal operations should be a distinct method